src/context.cpp

```diff
@@ -483,6 +484,102 @@ namespace librealsense
+
+        void enqueue_frame( T && frame, std::shared_ptr< rs2_stream_profile > prof )
+        {
+            std::lock_guard< std::mutex > lock( _queues_lock );
+            _frame_queue.push_back( std::move( frame ) );
+            _prof_queue.push_back( prof );
+            search_for_match(); //Call under lock
+        }
+
+        void enqueue_metadata( realdds::topics::flexible_msg && md )
+        {
+            std::lock_guard< std::mutex > lock( _queues_lock );
+            while( _metadata_queue.size() >= max_md_queue_size )
+                _metadata_queue.pop_front(); //throw oldest
+
+            _metadata_queue.push_back( std::move( md ) );
+            search_for_match(); //Call under lock
+        }
+
+        typedef std::function< void( T && ) > on_frame_ready;
+        typedef std::function< void( T & frame, const realdds::topics::flexible_msg & metadata ) > add_metadata_to_frame;
+
+        void reset( on_frame_ready cb = nullptr, add_metadata_to_frame add_md = nullptr )
+        {
+            std::lock_guard< std::mutex > lock( _queues_lock );
+            _frame_queue.clear();
+            _prof_queue.clear();
+            _metadata_queue.clear();
+            _cb = cb;
+            _add_md = add_md;
+        }
+
+    private:
+        //Call under lock
+        void search_for_match()
+        {
+            // Wait for frame + metadata set, but if metadata is lost pass frame to the user anyway
+            if( _frame_queue.empty() || ( _metadata_queue.empty() && _frame_queue.size() < max_frame_queue_size ) )
+                return;
+
+            // Sync using frame ID. Frame IDs are assumed to be increasing with time
+            int frame_id = _frame_queue.front().frame_number;
+            int md_id = _metadata_queue.empty() ? frame_id : // If no metadata create a "match"
+                        std::stoi( rsutils::json::get< std::string >( _metadata_queue.front().json_data(), "frame-id" ) );
+
+            while( frame_id > md_id && _metadata_queue.size() > 1 )
+            {
+                // Metadata without frame, remove it from queue and check next
+                _metadata_queue.pop_front();
+                md_id = std::stoi( rsutils::json::get< std::string >( _metadata_queue.front().json_data(), "frame-id" ) );
+            }
+
+            if( frame_id == md_id )
+                handle_match();
+        }
+
+        //Call under lock
+        void handle_match()
+        {
+            T & frame = _frame_queue.front();
+
+            if( !_metadata_queue.empty() && _add_md )
+            {
+                _add_md( frame, _metadata_queue.front() );
+                _metadata_queue.pop_front();
+            }
+
+            if( _cb )
+                _cb( std::move( frame ) );
+
+            _frame_queue.pop_front();
+            _prof_queue.pop_front();
+        }
+
+        on_frame_ready _cb = nullptr;
+        add_metadata_to_frame _add_md = nullptr;
```

Comment:
Isn't this pretty much the same as on-frame-ready but with metadata?
Why not pass the metadata to on-frame-ready (if we have it) and not resort to two different functions?

Reply:
This looked cleaner to me. With clearer names and definitions of funcitons, like `handle_video_data` and `add_video_frame_metadata`.
 If using one callback with a possible nullptr metadata parameter I would have something like `handle_video_frame_with_metadata` that part of it's logic has been handled in `handle_video_frame`.